### PR TITLE
Add a sharding rule for stablehlo BatchNorm family of ops

### PR DESCRIPTION
**Pull request by sgligorijevicTT**
### Ticket
closes https://github.com/tenstorrent/tt-xla/issues/3532

### Problem description
Shardy doesn't have a rule for batch norm training and grad ops

### What's changed
Added

### Checklist
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: lib/Dialect/StableHLO/Transforms/RegisterCustomShardingRule.cpp [hunk 1]
```diff
@@ -750,6 +750,58 @@ getAllToAllCombineShardingRule(mlir::stablehlo::CustomCallOp op) {
...
+    // Operands: input[dim], scale[?], bias[?]
+    // Results:  output[dim], mean[?], variance[?]
+    // 1-D operands/results participate only on the feature dimension (at dim
+    // 0).
+    int64_t oneDimIdx = (dim == featureIndex) ? 0 : sdy::kNullDim;
+    builder.addFactor({static_cast<int64_t>(dim), oneDimIdx, oneDimIdx},
+                      {static_cast<int64_t>(dim), oneDimIdx, oneDimIdx},
+                      dimSize,
+                      dim == featureIndex ? sdy::FactorType::kPassThrough
+                                          : sdy::FactorType::kNeedReplication);
```
[the remark this answers is not in the document]
Comment (sgligorijevicTT) [reply lib/Dialect/StableHLO/Transforms/RegisterCustomShardingRule.cpp:779]: spec guarantees tensor inputs, wontfix

File: lib/Dialect/StableHLO/Transforms/RegisterCustomShardingRule.cpp [hunk 2]
```diff
@@ -759,6 +811,12 @@ struct StablehloShardingModel
     if (auto scatterOp = llvm::dyn_cast<mlir::stablehlo::ScatterOp>(op)) {
       return getScatterShardingRule(scatterOp);
     }
+    if (auto bnOp = llvm::dyn_cast<mlir::stablehlo::BatchNormTrainingOp>(op)) {
+      return getBatchNormTrainingShardingRule(bnOp);
+    }
+    if (auto bnOp = llvm::dyn_cast<mlir::stablehlo::BatchNormGradOp>(op)) {
+      return getBatchNormGradShardingRule(bnOp);
+    }
```
[the remark this answers is not in the document]
Comment (sgligorijevicTT) [reply lib/Dialect/StableHLO/Transforms/RegisterCustomShardingRule.cpp:819]: added a test

File: test/ttmlir/Dialect/StableHLO/register_custom_sharding_rule/batch_norm.mlir [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+// RUN: ttmlir-opt --stablehlo-pipeline -o %t.mlir %s
+// RUN: FileCheck %s --input-file=%t.mlir
+
+// Test that batch_norm_training sharding rule correctly partitions the feature
+// dimension across all operands and results. The input tensor<2x6xbf16> is
+// sharded on dim 0 (batch) with _axis_0=2, then reshaped to tensor<1x12x1xbf16>
+// which merges batch into the feature dim. The sharding rule should propagate
+// _axis_0 through the feature dimension so that scale, bias, mean, and variance
+// (all 1-D) are partitioned consistently: 12 -> 6.
+module @batch_norm_training_feature_sharding attributes {mhlo.cross_program_prefetches = [], mhlo.input_output_alias = [], mhlo.is_dynamic = false, mhlo.use_auto_spmd_partitioning = false} {
```
Comment (sshonTT) [inline test/ttmlir/Dialect/StableHLO/register_custom_sharding_rule/batch_norm.mlir:8]: Could you also add a test case for `batchNormGrad`? Additionally, I’d like to see tests verifying that Shardy propagation correctly triggers ccl ops. For example, we should confirm that an all_gather is properly inserted when the scale tensor is sharded.
